Drop exception usage.

nicolasvasilache opened the pull request:
This revision uses llvm_unreachable instead.

andidr commented on CMakeLists.txt:
```diff
@@ -7,5 +7,6 @@ set(LLVM_ENABLE_PROJECTS "" CACHE STRING
   "Semicolon-separated list of projects to build, or \"all\". The mlir project is added by default.")
 list(APPEND LLVM_ENABLE_PROJECTS "mlir")
 
+link_directories(${CMAKE_CURRENT_BINARY_DIR}/llvm-project/llvm/lib)
```
Now obsolete after commit 04da3df.

andidr commented on teckyl/MLIRAffineExprGen.h:
```diff
@@ -69,7 +69,7 @@ class MLIRAffineExprGen {
         if (uintval >
             static_cast<uint64_t>(std::numeric_limits<int64_t>::max()))
-          throw Exception("Unsigned integer constant too big");
+          llvm_unreachable("Unsigned integer constant too big");
```
I think it would make sense to keep this exception. There is nothing fundamentally wrong with large integer constants and a caller might want to catch this case in order to provide more detailed feedback to the user. This could be wrapped in a macro that either evaluates to `llvm_unreachable` or that throws depending on a compiler flag.

andidr commented on teckyl/MLIRGen.cpp:
```diff
@@ -15,7 +11,13 @@
+#include "MLIRAffineExprGen.h"
+#include "lang_affine.h"
+#include "lang_extras.h"
+
```
These changes look like they came from clang-format with a different setting.

andidr commented on teckyl/MLIRGen.cpp:
```diff
@@ -310,7 +310,8 @@ mlir::Value buildBinaryExprFromValues(mlir::OpBuilder &builder, mlir::Value lhs,
 
-    throw mlirgen::SourceException(location, ss.str());
+    mlirgen::SourceException err(location, ss.str());
+    llvm_unreachable(err.what());
```
Same here: if this could be wrapped in a macro that would be good.

andidr commented on teckyl/MLIRGen.cpp:
```diff
@@ -320,8 +321,9 @@ mlir::Value buildBinaryExprFromValues(mlir::OpBuilder &builder, mlir::Value lhs,
-    throw mlirgen::SourceException(
+    mlirgen::SourceException err(
         location, "Cannot create binary operation: Unsupported operand type");
+    llvm_unreachable(err.what());
```
Same here.

andidr commented on teckyl/MLIRGen.cpp:
```diff
@@ -384,12 +386,12 @@ class MLIRValueExprGen : public MLIRGenBase {
 
       if (!(iss >> icst))
-        throw mlirgen::Exception("Could not build integer constant");
+        llvm_unreachable("Could not build integer constant");
```
Same here.

andidr commented on teckyl/MLIRGen.cpp:
```diff
@@ -476,7 +478,8 @@ class MLIRValueExprGen : public MLIRGenBase {
 
-      throw mlirgen::SourceException(location, ss.str());
+      mlirgen::SourceException err(location, ss.str());
+      llvm_unreachable(err.what());
```
Same here.

andidr commented on teckyl/MLIRGen.cpp:
```diff
@@ -504,7 +507,8 @@ class MLIRValueExprGen : public MLIRGenBase {
       std::cerr << ss.str() << std::endl;
-      throw mlirgen::SourceException(loc(t->range()), ss.str());
+      mlirgen::SourceException err(loc(t->range()), ss.str());
+      llvm_unreachable(err.what());
```
Same here.

andidr commented on teckyl/MLIRGen.cpp:
```diff
@@ -607,7 +611,8 @@ class MLIRGenImpl : protected MLIRGenBase {
 
-        throw mlirgen::SourceException(loc(param.range()), ss.str());
+        mlirgen::SourceException err(loc(param.range()), ss.str());
+        llvm_unreachable(err.what());
```
Same here.

andidr commented on teckyl/MLIRGen.cpp:
```diff
@@ -617,13 +622,11 @@ class MLIRGenImpl : protected MLIRGenBase {
              << "dimensions";
-
-          throw mlirgen::Exception(ss.str());
+          llvm_unreachable(ss.str().c_str());
```
Same here.

andidr commented on teckyl/MLIRGen.cpp:
```diff
@@ -858,7 +861,8 @@ class MLIRGenImpl : protected MLIRGenBase {
 
-      throw mlirgen::SourceException(loc(c.range()), ss.str());
+      mlirgen::SourceException err(loc(c.range()), ss.str());
+      llvm_unreachable(err.what());
```
Same here.

andidr commented on teckyl/MLIRGen.cpp:
```diff
@@ -991,7 +995,8 @@ class MLIRGenImpl : protected MLIRGenBase {
 
-        throw mlirgen::SourceException(loc(c.range()), ss.str());
+        mlirgen::SourceException err(loc(c.range()), ss.str());
+        llvm_unreachable(err.what());
```
Same here.

andidr commented on teckyl/MLIRGen.cpp:
```diff
@@ -1066,8 +1071,9 @@ class MLIRGenImpl : protected MLIRGenBase {
-                                     name);
+            mlirgen::Exception err("Multiple ranks found for output tensor " +
+                                   name);
+            llvm_unreachable(err.what());
```
Same here.

andidr commented on teckyl/tc/lang/lexer.cc:
```diff
@@ -17,6 +17,8 @@
 #include <cstring>
 
+#include <llvm/ADT/Twine.h>
+#include <llvm/Support/ErrorHandling.h>
```
I would prefer preserving the possible use of the lexer / parser / semantic analysis as library functions and see all exceptions preserved / wrapped in a preprocessor macro.

nicolasvasilache committed:
Drop exception usage by default

This revision uses llvm_unreachable instead where reasonable.
In other cases, the exception usage is wrapped behind a macro
and can be activated with the flag COMPILE_WITH_EXCEPTIONS.

Tested both with and without exceptions.

This revision also bumps the LLVM version to
b917d5e1a4f73365dc3661495bb99280d80a9551

nicolasvasilache commented:
Addressed, please let me know if you want other changes.

andidr commented on teckyl/Exception.h:
```diff
@@ -3,13 +3,20 @@
+#else
+#define THROW_OR_ASSERT(X) llvm_unreachable(X.what());
+#endif // COMPILE_WITH_EXCEPTIONS
+
```
Semicolon already provided by the caller

andidr commented on teckyl/MLIRGen.h:
```diff
@@ -14,12 +14,15 @@ namespace mlirgen {
   SourceException(const mlir::FileLineColLoc &l, const std::string &msg)
       : Exception(buildMessage(l, msg)) {}
+  
+  const char* what() { return getMessage().c_str(); }
```
Same here

andidr commented on teckyl/MLIRGen.h:
```diff
@@ -14,12 +14,15 @@ namespace mlirgen {
 class Exception : public teckyl::Exception {
 public:
   Exception(const std::string &msg) : teckyl::Exception(msg) {}
+  const char* what() { return getMessage().c_str(); }
```
Already defined in parent class

andidr commented on teckyl/main.cc:
```diff
@@ -97,24 +94,14 @@ int main(int argc, char **argv) {
     default:
-      throw teckyl::Exception("Unknown action");
-    }
-  } catch(teckyl::Exception& e) {
```
This breaks exception handling when `COMPILE_WITH_EXCEPTIONS` is enabled

nicolasvasilache replied:
Fixed, thanks! I went for a bit more spaghetti #ifdef than I'd like but the tradeoff seemed to be to duplicate code.
If this grows in the future, it's likely more cases will be added in the switched so I optimized for minimizing future code duplication.

andidr commented on teckyl/tc/lang/error_report.h:
```diff
@@ -19,15 +19,25 @@
 #include "tc/utils/compiler_options.h"
 
+#ifdef COMPILE_WITH_EXCEPTIONS
+#define THROW_OR_ASSERT(X) throw X;
```
Same comment as above

nicolasvasilache commented:
Addressed.
